### PR TITLE
[WIP] Tracing: Update search API to return dataframes, support field extraction

### PR DESCRIPTION
**Pull request by dbczumar**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/dbczumar/mlflow/pull/11797?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11797/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11797
```

</p>
</details>

### What changes are proposed in this pull request?

Update search API to return dataframes, support field extraction

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracing/fluent.py [hunk 1]
```diff
@@ -209,11 +219,12 @@ def get_traces(n: int = 1) -> List[Trace]:
...
 
 def search_traces(
-    experiment_ids: List[str],
+    experiment_ids: Optional[List[str]] = None,
     filter_string: Optional[str] = None,
     max_results: Optional[int] = None,
     order_by: Optional[List[str]] = None,
-):
+    extract_fields: Optional[List[str]] = None,
+) -> "pandas.DataFrame":
```
Comment (harupy) [inline mlflow/tracing/fluent.py:227]: Does `search_traces` have to work in mlflow-skinny (which doesn't include pandas as a dependency)?

Comment (harupy) [reply mlflow/tracing/fluent.py:227]: We don't any rule or policy for what must work in mlflow-skinny tho.

Comment (jessechancy) [reply mlflow/tracing/fluent.py:227]: Do you think you can make the call here? @harupy

Comment (BenWilson2) [reply mlflow/tracing/fluent.py:227]: Let's add a try/catch around the import from within this API to make it clear that this API requires pandas to return a DF from the fluent API

File: mlflow/tracing/__init__.py
Comment (B-Step62) [inline mlflow/tracing/__init__.py:1]: nit: We may want to have it in utils or seperate module like `dataframe_utils.py`. Historically `__init__.py` file tends to become huge dump.

File: mlflow/tracing/__init__.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.tracing.utils import SPANS_COLUMN_NAME, traces_to_df
+
+if TYPE_CHECKING:
+    import pandas as pd
+
+    import mlflow.entities
+
+
+def extract(
```
Comment (B-Step62) [inline mlflow/tracing/__init__.py:14]: nit: This only seems to be used by `search_traces()` method, can we move it under `fluent.py` and private?

File: mlflow/tracing/utils.py [hunk 3]
```diff
@@ -101,3 +109,53 @@ def build_otel_context(trace_id: int, span_id: int) -> trace_api.SpanContext:
...
+            tags=trace.info.tags,
+            response=trace.data.response,
+        )
+        for trace in traces
+    ]
+    return pd.DataFrame.from_records([row.to_dict() for row in rows])
+
+
+@dataclass
+class _TraceRow:
```
Comment (B-Step62) [inline mlflow/tracing/utils.py:139]: Do we need this data class? It seems the dataclass is constructed and converted to dictionary immediately. Would it be simpler to define `to_pandas_dataframe_row()` method in the `Trace` dataclass?

File: tests/tracing/test_fluent.py [hunk 4]
```diff
@@ -281,3 +282,292 @@ def predict(self, x, y):
...
+    monkeypatch.setattr("mlflow.tracing.fluent.MlflowClient", MockMlflowClient)
+
+    df = mlflow.search_traces()
+    assert df["response"].isnull().all()
+    assert df["tags"].tolist() == [{}]
+    assert df["request_metadata"].tolist() == [{}]
+
+
+def test_search_traces_extracts_fields_as_expected(monkeypatch):
+    class TestModel:
```
Comment (B-Step62) [inline tests/tracing/test_fluent.py:349]: nit: This test model is defined many times, let's define outside the cases and reuse🙂

**Pull request closed without merging**